examples/Example2/example2.cu

```diff
@@ -0,0 +1,224 @@
+// SPDX-FileCopyrightText: 2020 CERN
+// SPDX-License-Identifier: Apache-2.0
+
+#include "example2.h"
+
+#include "process.h"
+#include "process_list.h"
+#include "pair_production.h"
+#include "energy_loss.h"
+
+#include "track.h"
+
+#include <AdePT/BlockData.h>
+#include <AdePT/LoopNavigator.h>
+#include <AdePT/MParray.h>
+
+#include <VecGeom/base/Config.h>
+#ifdef VECGEOM_ENABLE_CUDA
+#include <VecGeom/backend/cuda/Interface.h>
+#endif
+
+#include <curand.h>
+#include <curand_kernel.h>
+
+#include <iostream>
+#include <iomanip>
+#include <stdio.h>
+
+// some simple scoring
+struct Scoring {
+  adept::Atomic_t<int> hits;
+  adept::Atomic_t<int> secondaries;
+  adept::Atomic_t<float> totalEnergyLoss;
+
+  VECCORE_ATT_HOST_DEVICE
+  Scoring() {}
+
+  VECCORE_ATT_HOST_DEVICE
+  static Scoring *MakeInstanceAt(void *addr)
+  {
+    Scoring *obj = new (addr) Scoring();
+    return obj;
+  }
+};
+
+constexpr double kPush = 1.e-8;
+
+// kernel select processes based on interaction lenght and put particles in the appropriate queues
+__global__ void DefinePhysicalStepLength(adept::BlockData<track> *block, process_list *proclist,
+                                         adept::MParray **queues, Scoring *scor)
+{
+  int n = block->GetNused() + block->GetNholes();
+
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < n; i += blockDim.x * gridDim.x) {
+    // skip particles that are already dead
+    track &mytrack = (*block)[i];
+    if (mytrack.status == dead) continue;
+
+    // return value (if step limited by physics or geometry) not used for the moment
+    // now, I know which process wins, so I add the particle to the appropriate queue
+    float physics_step = proclist->GetPhysicsInteractionLength(i, block);
+    float step         = LoopNavigator::ComputeStepAndPropagatedState(mytrack.pos, mytrack.dir, physics_step,
```

Comment:
Indeed, the geometry step has to be computed after having the physics step limitation. In a future version we have to try to split it out as separate kernel, i.e. cache the physics proposed step in the track and exit this, then pass the relay to the geometry kernel. We can do it later though.